### ISSUE TITLE
Do pre-decompress gzip with IAA hardware 

### ISSUE THREAD

**Issue by yaqi-zhao** [facebookincubator/velox#5718]
### Description

The Intel® In-Memory Analytics Accelerator (Intel® IAA) is a hardware accelerator that provides very high
throughput compression and decompression combined with primitive analytic functions. It is available in the newest generation of Intel® Xeon® Scalable processors ("Sapphire Rapids"). 
We can offload the GZip decompression to the IAA hardware and save the CPU bandwidth. Here is a description of how to offload the GZip decompression to the IAA hardware.
<img width="679" alt="image" src="https://github.com/facebookincubator/velox/assets/100556669/d3ec7364-a964-4536-a838-089ec88d1651">
We use IAA to do pre-decompress work in Velox native parquet reader and the get a good performance in the TPCH Benchmark. The TPCH E2E results show that:
1) About 2X performance gain compared with the current gzip solution
2) About 20~40% performance gain compared with ZSTD compressed file
3) About 5~10% performance compared with snappy compressed file.

**Comment by pedroerp** [comment]
Cc: @oerling @Yuhta

**Comment by yaqi-zhao** [comment]
Hi, @Yuhta @oerling, we have completed the development of the solution as description. Do you think it is reasonable to submit a PR for your review?

**Comment by george-gu-2021** [comment]
Regarding checking "whether IAA can decompress the data", the logic is like that:
step-1,  detect whether the system includes usable IAA Hardware? If yes, go to step-2;
step-2,  detect whether the compression format is zlib? If yes, go to step-3;
step-3,  read the window size info from zlib header, and detect whether windows size is <= 4KB? If yes, go to IAA HW decompress;

Otherwise, fallback to the Velox SW path to decompress the data.

Hi @yaqi-zhao , be free add your comments if the above description is inaccurate. Thanks!

**Comment by Yuhta** [comment]
I am curious how you implement `readWithVisitor`.  Ideally we should not duplicate any of the logic already in `PageReader`.

**Comment by yaqi-zhao** [comment]
Hi, @Yuhta  Actually to avoid affect the current <PageReader> file, the new added <IAAPageReader> has some duplicate code. <readWithVisitor> of <IAAPageReader> is the same logic as <PageReader>.  You can see the implementation at https://github.com/facebookincubator/velox/pull/6176/files#diff-3155c9fa38e02bffd7c4bf31b1eed7a39e2f12efdae4b917817eabb388cc9221. 
Maybe I can move the logic of <IAAPageReader> to <PageReader> if it is not allowed to duplicate the logic of <PageReader>.

**Comment by Yuhta** [comment]
@yaqi-zhao Yes if you can implement the logic inside `PageReader` (with a few conditional includes) that would be best.

**Comment by Yuhta** [comment]
@yaqi-zhao Also I would recommend rebase your work on #5914 , so that `uncompressQplData` can be added to common compression code (maybe with a better async interface)

**Comment by yaqi-zhao** [comment]
Hi, @Yuhta. I rebase my work on #5914. ```IAAPageReader``` is deleted and the new logic is implemented inside the ```PageReader```.   The change is updated at the #6176 .
